packages/block-editor/src/utils/dom.js

```diff
@@ -57,3 +57,132 @@ export function getBlockClientId( node ) {
 
 	return blockNode.id.slice( 'block-'.length );
 }
+
+/*
+ * Cache for rectUnion() return values.
+ * When popovers are present on the canvas,
+ * this calculation function is called
+ * multiple times with the same values.
+ */
+const rectUnionCache = new Map();
```

Comment:
This is necessary (for now) because, if not cached, `rectUnion()` is called a few times every second.

On a page left untouched with a popover open, the call count can potentially reach `n`.

This cache reduces the times the internal logic of `rectUnion` runs dramatically.

Most of the functions up the chain that use this function are written with `useCallback` and `useMemo`, but it  indicates that the Popover component is updating all these values using `requestAnimationFrame`.

Presumably indirectly, via `useFloating` and/or `autoUpdate` from `floating-ui/@floating-ui/react-dom

https://github.com/WordPress/gutenberg/blob/d712323418d5c0a41983aea18060330d9773fbf9/packages/components/src/popover/index.tsx#L291-L291

Reply:
One possible problem would be that this cache is never cleared. So a long-running page with multiple popovers might cause memory leaks.

What if we use memoization rather than a global cache? [`memize`](https://github.com/aduth/memize) might be helpful here. We can also create and clear the memoization for each instance. This should also help us avoid the need to generate a cache key string using `JSON.stringify`.

Reply:
Great idea, thanks for the feedback

Reply:
I tried a few rounds with memoizing the function, and it made things a little janky when scrolling the page, even with a `maxSize` to limit the cache size.

So I've taken out the optimization for now. It wasn't in response to any perceived performance issue, only a defensive tactic.

Reply:
Is it possible that you're using the `DOMRect` as the input parameter which cannot be memoized? I think the simplest way is to just flatten the input parameters and use primitives like so:

```js
rectUnion(leftX, leftY, leftWidth, leftHeight, rightX, rightY, rightWidth, rightHeight)
```

It gets a little verbose but then they're all primitives which can be memoized accordingly.

That said, we don't need to do this at all if this is not the bottleneck until proven otherwise! 👍

Reply:
>  we don't need to do this at all if this is not the bottleneck until proven otherwise

yeah, there's not demonstrated bottleneck, I was just surprised that this function was running 100,000s of times in a short period of time

Reply:
Caching comes with its own performance concerns 😀 So long as `isElementVisible` is fast it should be okay. I ordered the checks in order of most performant to least performant.